### ISSUE TITLE
Parallel build of different projects

### ISSUE THREAD

**Issue by pierricgimmig** [liamkf/msfastbuild#7]
For solutions with multiple projects, every project is treated sequentially.  Could obj's from different projects be compiled in parallel!?

**Comment by jairbubbles** [comment]
Could not agree more with that one. I think the build is slower because of that. I'll make some comparison tests as soon  as I can compile my  more complex projects which already have .bff generated by another project generator.
But the other issue is that when you have FastBuild Monitor you only see the last project which is most of the time pretty useless on a multi-projects solution.

**Comment by liamkf** [comment]
I've been thinking about this and I don't think it'll be feasible until I change it to generate a bff for the whole solution...

**Comment by jairbubbles** [comment]
Yes, it seems so. We generate one bff with our project generator which includes all the other .bff I think.
It has some drawbacks though, if you change a small flag in one project you will recompile everything.

**Comment by TheClassic** [comment]
We're looking at using fastbuild and would need project parallelism. I'm interested in contributing. I don't have a lot of free time, but I'll definitely be looking into this. @jairbubbles , are you able to share your bff generator?

**Comment by jairbubbles** [comment]
Hello. Our bff generator is used internally and is not available for everyone. Anyway it's quite similar to premake but in C# so you need to describe your projects in script files prior to generating bff files.

**Comment by TheClassic** [comment]
Does anybody have a good test case for this. I haven't resolved all the issues that prevent my solution from building successfully yet.

I plan on starting on this tonight and seeing if I can knock it out this weekend. Feel free to assign to me.

**Comment by TheClassic** [comment]
I have a simple case here that I think represents more or less what we want. Let me know what you think:
https://github.com/TheClassic/msfastbuildTests/tree/81abe71a6d3330dd4554959184c1dfa53f7f3e9c/projectReferences

**Comment by TheClassic** [comment]
I've made some progress with this, I'd say I'm about 1/3 of the way there. The first step was to refactor a bit. I think it makes sense to discuss this approach at this stage with the stakeholders. 

This shows the changes:
https://github.com/liamkf/msfastbuild/compare/master...TheClassic:parallel_projects?expand=1
Run time is unaffected at 35 seconds on my specs for a solution with 268 projects and some hefty dependency trees. I can imagine memory usage has gone up. I need to test to make sure no functionality was broken.

Project references are evaluated recursively as before, but bff files are generated as the loop unrolls. The changes effectively invert the approach, dependencies are tracked instead of dependents. These changes will allow a bff file to include bff files for dependencies and reference symbols defined in them.

I'm curious about your thoughts on this approach.
